### PR TITLE
Advanced Building Blocks : Enumerables

### PR DESCRIPTION
**Pull request by DeliceLydia**
In this project, we implemented our own version of Ruby methods 

This is the list of methods we implemented:

- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject

We used as well:

- Yield
- Call method

### CODE REVIEW

File: enumerables.rb
Comment (peterrobert) [inline enumerables.rb]: - [x] My_each should return an enumerator if there is no block given. Kindly refer [here](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each);

### YOURS
yours no block

### EXPECTED
expected no block

- [x] When a block is given when self is an array it should return the array itself after calling the given block once.
### yours

yours block given

### expected

expected block given

- [x] Kindly make sure it works for all other classes like a Hash etc. It isn't working and the same errors as given above :arrow_double_up:  appears.

File: enumerables.rb
Comment (peterrobert) [inline enumerables.rb]: - [ ] It returns an enumerator if no block is given. Just like with the my_each above the same case applies to my each with index.

- [ ]  when a block is given when self is an array returns the array itself after calling the given block once for each element in self.

- [ ] Also make sure it works for all other classes like a hash and a range etc

Kindly refer [refer](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index)

**Review by peterrobert — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED :recycle: :stop_sign: 
### HELLO :wave: @DeliceLydia  AND @TSHEPO-CLOUD . GOOD PROGRESS ON THE PROJECT. YOU ARE IN THE RIGHT DIRECTION. HOWEVER, THERE are SOME CHANGES THAT NEED TO BE DONE. 

### kindly check below :arrow_down: 

- [ ] Kindly double-check all your methods. Try and follow this guide [here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index);

- [ ] Also I can see you have the linter files added. Am guessing you forgot to enable the linter in your repo kindly do so.

### HINT 

- [ ] Remember since you are re-using some of your methods like my_each and my_each_with_index if they don't work, none of the other methods will work. Hence this is one of the reasons the methods aren't working. Kindly make sure these methods do work before you proceed.


### if you have any question concerning the review, reach me on slack @peterrobert

### HAPPY CODING :computer:

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: - when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- when a block is given when self is a hash behaves like #each calls the given block once for each element in self

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: - Same as my_each

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: - returns an array containing all elements of enum for which the given block returns a true value, kindly fix it to make it work with a range.
TIP: you can convert it into an array

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: - returns false if the block returns false or nil
- when no block or argument is given returns true when none of the collection members are false or nil
- when no block or argument is given returns false when one of the collection members are false or nil
- when a class is passed as an argument returns true if all of the collection is a member of such class
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: - returns true if the block ever returns a value other than false or nil
- returns false if the block ever returns a value false or nil
- when no block or argument is given returns true if at least one of the collection is not false or nil
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: - returns false if the block ever returns true for all elements
- when no block or argument is given returns false only if one of the collection members is true
- when a class is passed as an argument returns true if none of the collection is a member of such class
- when a Regex is passed as an argument returns false if any of the collection matches the Regex
- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: - counts the number of elements yielding a true value if a block is given
- does not mutate the original array

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: - returns a new array with the results of running a given block once for every element in enum
- returns a new array with the results of running a given block once for every element in enum
- returns an Enumerator if no block is given
- accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: - searches for the longest word in an array of strings
- when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block
- when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED ♻️ 

Hello team!
Good job with the changes made, but still have to implement some to get the project approved 💪 
Kindly, check my comments below 👇  apply the changes and resubmit your project for a new code review

Good luck 😃 

## Mandatory changes

File: enumerables.rb
Comment (ghost) [inline enumerables.rb:18]: Your my_each method should also iterate over a range of values and hash values.

-  `each` for range values

```
a = (0..3)
a.each {|x| print x}
result: 0123
```

-  `my_each` for range values

```
a = (0..3)
a.my_each {|x| print x}
result: enumerables.rb:7:in `my_each': undefined local variable or method `length' for 0..3:Range (NameError)
```

- `each` for hash values.

```
h = {foo: 0, bar: 1, baz: 2}
h.each {|key, value| print "#{key}: #{value}"}

result: foo: 0bar: 1baz: 2
```

- `my_each` for hash values.

```
h = {foo: 0, bar: 1, baz: 2}
h.my_each {|key, value| print "#{key}: #{value}"}

result: : : :
```

File: enumerables.rb
Comment (ghost) [inline enumerables.rb:28]: Your my_each_with_index should iterate over a range of values and hash values.

- `each_with_index` for range values.

```
hash = Hash.new
(0..5).each_with_index { |item, index|
  hash[item] = index
}
print hash  #=> {0=>0, 1=>1, 2=>2, 3=>3, 4=>4, 5=>5}
```

- `my_each_with_index` for range values.
```
hash = Hash.new
(0..5).my_each_with_index { |item, index|
  hash[item] = index
}
print hash   #=> enumerables.rb:14:in `my_each_with_index': undefined local variable or method `length' for 0..5:Range (NameError)
``` 

- `each_with_index` for hash values.

```
hash = Hash.new
{foo: 0, bar: 1, baz: 2}.each_with_index { |item, index|
  hash[item] = index
}
print hash  #=> {[:foo, 0]=>0, [:bar, 1]=>1, [:baz, 2]=>2}
```

- `my_each_with_index` for hash values.
```
hash = Hash.new
{foo: 0, bar: 1, baz: 2}.my_each_with_index { |item, index|
  hash[item] = index
}
print hash   #=> {nil=>2}
```

File: enumerables.rb
Comment (ghost) [inline enumerables.rb:39]: my_select should iterate over a range of values.

`print (0..2).select { |num|  num.even?  } 
#=> [0, 2]
`
`print (0..2).my_select { |num|  num.even?  } 
#=> enumerables.rb:7:in `my_each': undefined local variable or method `length' for 0..2:Range (NameError)
`

File: enumerables.rb
Comment (ghost) [inline enumerables.rb]: Fix `my_all?` for the following:

1). **In the case below, `my_all?` should return true if every element in the array contains letter /t/.**
```
%w[ant bear cat].all?(/t/)   #=> false
%w[ant bear cat].my_all?(/t/)  #=> enumerables.rb:26:in `my_all?': wrong number of arguments (given 1, expected 0) (ArgumentError)
```
2). **`my_all` checks if all values are numeric.**
```
print [1, 2i, 3.14].all?(Numeric) #=> true
print [1, 2i, 3.14].my_all?(Numeric) #=> enumerables.rb:26:in `my_all?': wrong number of arguments (given 1, expected 0) (ArgumentError)
```

3).   **If no block is given, it should return true if none of the elements in the array is false or nil.**
```
print [nil, true, 99].all?  #=> false
print [nil, true, 99].my_all?  #=> enumerables.rb:28:in `block in my_all?': no block given (yield) (LocalJumpError)
```
4). **`my_all` should iterate over range values**
```
print (0..2).all?(Numeric)   #=> true
print (0..2).my_all?(Numeric)   #=> enumerables.rb:26:in `my_all?': wrong number of arguments (given 1, expected 0) (ArgumentError)
```

File: enumerables.rb
Comment (ghost) [inline enumerables.rb]: Fix the following for `my_any?` method.

1). **Returns true if any element has a length of 3.**
```
%w[ant bear cat].any? { |word| word.length >= 3 } #=> true
%w[ant bear cat].my_any? { |word| word.length >= 3 } #=> false
```
2). **Returns true if any element in the array has a letter /d/.**
```
print %w[ant bear cat].any?(/d/)  #=> false
print %w[ant bear cat].my_any?(/d/)  #=> enumerables.rb:32:in `my_any?': wrong number of arguments (given 1, expected 0) (ArgumentError)
```
3). **Should return true if any element in the array is an integer.**
```
print [nil, true, 99].any?(Integer) #=> true
print [nil, true, 99].my_any?(Integer) #=> enumerables.rb:32:in `my_any?': wrong number of arguments (given 1, expected 0) (ArgumentError)
```
4). **Should return true if any of the value in the array is a truthy.**
```
print [nil, true, 99].any? #=> true
print [nil, true, 99].my_any? #=> enumerables.rb:34:in `block in my_any?': no block given (yield) (LocalJumpError)
```

5). **Should iterate over a range of values.**
```
print (0..2).any? #=> true
print (0..2).my_any? #=> enumerables.rb:7:in `my_each': undefined local variable or method `length' for 0..2:Range (NameError)
```

File: enumerables.rb
Comment (ghost) [inline enumerables.rb]: Fix the following for `my_none?` method.

1). **Returns true if none of the elements in the array has a letter /d/.**
```
print %w{ant bear cat}.none?(/d/) #=> true
print %w{ant bear cat}.my_none?(/d/) #=> enumerables.rb:38:in `my_none?': wrong number of arguments (given 1, expected 0) (ArgumentError)
```
2). **Returns true if none of the elements in the array belongs to the Float class.  A float is a number that contains a decimal point.**
```
print [1, 3.14, 42].none?(Float) #=> false
print [1, 3.14, 42].my_none?(Float) #=> enumerables.rb:38:in `my_none?': wrong number of arguments (given 1, expected 0) (ArgumentError)
```
3). **Returns true if all elements in the array are falsey.**
```
print [nil, false].none? #=> true
print [nil, false].my_none? #=> enumerables.rb:40:in `block in my_none?': no block given (yield) (LocalJumpError)
```
4). **Should iterate over a range of values.**
```
print (0..2).none? #=> false
print (0..2).my_none? #=> enumerables.rb:7:in `my_each': undefined local variable or method `length' for 0..2:Range (NameError)
```

File: enumerables.rb
Comment (ghost) [inline enumerables.rb:169]: Fix the following for `my_count` method

1). **It should return the length of the array if no argument or block is given.**
```
print [1, 2, 4, 2].count   #=> 4
print [1, 2, 4, 2].my_count   #=> enumerables.rb:44:in `my_count': wrong number of arguments (given 0, expected 1) (ArgumentError)
```
2). **If a block is given, it counts the number of elements yielding a true value.**
```
print [1, 2, 4, 2].count{ |x| x%2==0 } #=> 3
print [1, 2, 4, 2].my_count{ |x| x%2==0 }  #=>enumerables.rb:44:in `my_count': wrong number of arguments (given 0, expected 1) (ArgumentError)
```
3). **It should iterate over a range of values.**
```
print (1..5).count{ |x| x%2==0 } #=> 2
print (1..5).my_count{ |x| x%2==0 }  #=> enumerables.rb:44:in `my_count': wrong number of arguments (given 0, expected 1) (ArgumentError)
```

File: enumerables.rb
Comment (ghost) [inline enumerables.rb]: Fix the following for `my_map` method.

1). **It should iterate over a range of values.**
```
print (1..4).map { |i| i*i }   #=> [1, 4, 9, 16]
print (1..4).my_map { |i| i*i }   #=> enumerables.rb:7:in `my_each': undefined local variable or method `length' for 1..4:Range (NameError)
```
2). **It should print an enumerator if not block is given.**
```
print [1,2,4,3].map  #=> #<Enumerator:0x00005579dd9b6070>
print [1,2,4,3].my_map  #=> enumerables.rb:53:in `block in my_map': no block given (yield) (LocalJumpError)
```

File: enumerables.rb
Comment (ghost) [inline enumerables.rb]: Fix the following for `my_inject` method.

1). **It should iterate over a range of values.**
```
print (5..10).inject { |sum, n| sum + n } #=> 45
print (5..10).my_inject { |sum, n| sum + n }  #=> enumerables.rb:7:in `my_each': undefined local variable or method `length' for 5..10:Range (NameError)
```
2). **`my_inject `method returns zero when a block is given and no argument is given. That should not be the case.**
```
print [1,2,3,4].my_inject(1) { |product, n| product * n }  #=> 24
print [1,2,3,4].my_inject { |product, n| product * n } #=> 0
```

`inject` method gives the following results
```
print [1,2,3,4].inject(1) { |product, n| product * n }  #=> 24
print [1,2,3,4].inject { |product, n| product * n } #=> 24
```

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

## Hello team,

**Great job on your project. However, there are a few changes still needed before your project can be approved.**

## Mandatory
1). **There is a rubocop case that is still failing after I run `rubocop -a`.**

rubocop

2). **Please check the comments below. I used [ruby docs](https://ruby-doc.org/core-2.6.5/Enumerable.html) provided on the pathwright for reference.**

### Happy coding 😃.

**PS: You can reach me on slack (Rose Wanjohi).**

File: enumerables.rb
Comment (peterrobert) [inline enumerables.rb]: ### There are still some changes to be made in this method.

- [ ] When you pass a class as an argument for example let's say a class of numeric(numbers), the method should return true if all the members in the collection are of such a class.
### YOURS
numeric yours
### ORIGINAL
numeric  original

File: enumerables.rb
Comment (peterrobert) [inline enumerables.rb]: ### There are also some changes required here.

- [ ] When both a proc and a block is given, this method should only execute the proc and returns a new array with the results of running the proc for every element.

### HINT 

- A conditional statement will do the trick here.

File: enumerables.rb
Comment (peterrobert) [inline enumerables.rb]: ### SOME CHANGES REQUIRED
- [ ] Kindly you need to restructure this method it is failing for the most part. I would recommend that you use [this link](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject) and try and restructure it again

**Review by peterrobert — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED :recycle: :stop_sign: 
### HELLO :wave: @DeliceLydia NICE PROGRESS. YOU HAVE MADE  MOST OF THE CHANGES REQUESTED BY THE PREVIOUS TSE. JUST SOME FEW MORE CHANGES TO BE DONE. KINDLY CHECK BELOW :arrow_down: 

- [ ] There are still some rubocop offenses. Kindly fix all rubocop offenses before you submit for another review.

rubocop


### if you have any question, kindly reach me on slack @peterrobert 

### HAPPY CODING :computer:

**Review by Rhelli — CHANGES_REQUESTED** [review]
# Changes Required 🛑 ♻️

Hi guys, well done on the project so far and making the amendments proposed by the previous TSEs. There are however still some tests failing on some of your methods, so lets get them fixed and get you on to the next project! We have a total of 2 test failures here, so I'll list and detail them below.

******
### 1. **[Mandatory]** - my_map should return an enumerator if no block is given.

image

**Issue:** The issue here is self explanatory. I would, however look to see if your my_each is functioning correctly and address the errors traceback (blue text in the error) as it appears to route back to your `my_each` method. If your `my_each` doesnt function correctly, it will affect a lot of the rest of your code here.

********

### 2. **[Mandatory]** - my_map should execute only the proc when **both** a block and a proc are given and should return a new array with the results

image


**Issue:** The issue appears to be the way in which you are calling your proc. Please address this.

*****

### 3. **[Mandatory]** - Please fix the incorrect indentation current present in your `my_inject` and `my_map` methods.

image


image


********

Get these issues sorted and you should be bug free and ont the next project in no time. Remember - your `my_each` method has to be rock solid for this project, so ensure it is working 💯! Good luck!

All the best,

Rory

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: Check these lines of code! You already know there is a block given, maybe you should check the state of the proc.... 🤓

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUESTED ♻️ 

Great job team! I can see you put a lot of effort into this project!  👏 
But we still need to implement some changes to get it approved 💪 
Kindly fix the issues and resubmit your project for a new code review

## Mandatory changes
- my_map enumerable
   - my_map should not mutate the original array
   - #my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.
   - #my_map returns a new array with the results of running a given block once for every element in a range
   - #my_map returns a new array with the results of running a given block once for every element

**Comment by Rhelli** [comment]
# Approved! ✅ 🚀

@DeliceLydia, @TSHEPO-CLOUD, congratulations and excellent work on fixing the changes proposed by the previous TSEs! I can happily say that your project is passing all tests and you have the green light!

image

Give yourselves a pat on the back and move on to the next project! 👏🏽  Good luck!

All the best,

💂🏽 Rory

**Pull request merged**
